src/cockpit/channels/filesystem.py

```diff
@@ -172,11 +211,19 @@ def delete(self, path: str, tag: 'str | None') -> str:
         return '-'
 
     async def set_contents(
-        self, path: str, tag: 'str | None', data: 'bytes | None', size: 'int | None'
+        self, path: str, tag: 'str | None', data: 'bytes | None', size: 'int | None',
+        attrs: 'FSReplaceAttrs | None'
     ) -> 'str | None':
         dirname, basename = os.path.split(path)
         tmpname: str | None
         fd, tmpname = tempfile.mkstemp(dir=dirname, prefix=f'.{basename}-')
+
+        def chown_if_required(fd: 'int', buf: 'os.stat_result | None' = None):
+            if attrs is None and buf is not None:
+                os.fchown(fd, buf.st_uid, buf.st_gid)
+            elif attrs is not None and attrs.uid is not None and attrs.gid is not None:
+                os.fchown(fd, attrs.uid, attrs.gid)
```

Comment:
This is over-complicated.  I think you mean to say:

```
if attrs is not None and attrs.gid is not None and attrs.uid is not None:
   ...attrs way...
elif buf is not None:
   ...buf way...
```

The current logic looks like it would be impacted by attrs being provided but being empty and I'm not sure that's intended?  If so, it definitely needs a comment.